### ISSUE TITLE
Implement Gas-Aware Routing with Cost Function Optimization

### ISSUE THREAD

**Issue by 0xtechroot** [0xtechroot/cowSolver#68]
## Overview
Implement gas-aware routing that minimizes total cost: `economic_loss + λ * estimated_gas`, achieving 30-50% gas savings.

### 1. Gas-Aware Cost Function
**Multi-objective optimization:**
```rust
fn total_cost(route: &Route, λ_gas: f64) -> f64 {
    let economic_loss = route.price_impact + route.fees;
    let gas_cost = estimate_gas(route);
    
    economic_loss + λ_gas * gas_cost
}
```

**Penalize:**
- Deep multi-hop routes
- Complex AMM interactions
- Heavy SLOAD/SSTORE operations
- Complex callbacks

**Favor:**
- Single-hop where price impact is small
- Simple math AMMs
- Pre-deployed router contracts
- Shared logic across trades

#### Batch Settlement
- One transaction settles N orders
- Amortize fixed costs
- Shared verification logic

#### Compact Calldata
```solidity
// Pack fields efficiently
struct CompactOrder {
    uint128 amount;      // Instead of uint256
    uint64 deadline;     // Instead of uint256
    uint32 poolId;       // Instead of address
    bytes32 signature;   // Compressed
}
```

**Techniques:**
- Use uint128/uint64 where possible
- Pack into uint256/bytes32
- Minimize dynamic arrays
- Fixed-size where feasible

#### Storage Minimization
**Stateless settlement:**
- Most state lives off-chain
- On-chain stores only:
  - User balances/vaults
  - Global configuration
  - Hashes/commitments
- Pass full state deltas
- Verify via hashes/signatures

### 3. Signature Optimization
**EIP-712 structured data:**
- Aggregate user intents off-chain
- Single multi-intent verification
- Shared parameters
- Amortize crypto operations

### 4. Gas Model Integration
**Light-weight gas prediction:**
```rust
struct GasModel {
    base_cost: u64,
    per_hop_cost: u64,
    storage_read: u64,
    storage_write: u64,
    external_call: u64,
}

fn estimate_route_gas(route: &Route, model: &GasModel) -> u64 {
    model.base_cost +
    route.hops.len() as u64 * model.per_hop_cost +
    route.storage_ops.reads * model.storage_read +
    route.storage_ops.writes * model.storage_write +
    route.external_calls * model.external_call
}
```

#### Balancer V2 Vault
- Batch swaps net all movements
- Single settlement transaction
- Internal balance reuse
- Relayer pattern

#### Uniswap V3
- Universal Router for multi-hop
- Minimal state touching
- Efficient tick traversal

### 6. Route Pruning
**Gas-aware pathfinding:**
- Prune routes with high gas overhead
- Minimal-hop preference
- Minimal-state-touching solutions
- Early termination for expensive paths

## Implementation Phases

**Phase 1:** Gas model development
- Per-operation cost database
- Per-protocol overhead tracking
- Historical gas analysis
- Dynamic fee estimation

**Phase 2:** Cost function integration
- Multi-objective routing
- Gas cost in scoring
- User-configurable λ_gas
- Route comparison with gas

**Phase 3:** On-chain optimization
- Calldata compression
- Storage pattern optimization
- Batch settlement
- Signature aggregation

**Phase 4:** Continuous optimization
- Gas profiling per route
- A/B testing
- Regression detection
- Automatic tuning

## Performance Targets
- **30-50%** gas savings vs naive routing
- **<5%** economic loss from gas optimization
- **Sub-millisecond** gas estimation
- **99%+** gas estimate accuracy

## Monitoring Metrics
- Actual vs estimated gas
- Gas cost per route type
- Gas savings vs baseline
- Economic loss from gas optimization

## Safety Considerations
- Never sacrifice security for gas
- Maintain slippage protection
- Verify all optimizations
- Fallback to safe paths

## References
- EIP-1559 gas mechanics
- Balancer V2 gas optimization
- Uniswap V3 router patterns
- CoW Protocol settlement optimization

## Priority
**High** - Significant cost reduction for users

## Labels
`enhancement`, `gas-optimization`, `performance`, `research`
